File: src/levels/meta/level05.ts
import type { LevelMeta } from '../../lib/types';

const meta: LevelMeta = {
  id: 5,
  slug: 'the-drunkards-drift',
  title: "The Drunkard's Drift",
  tier: 'FOUNDATIONS',
  tagline: 'the first process with memory of its mistakes',
  brief: String.raw`Position log of an autonomous unit whose navigation drifts: 1000 steps of

$$x_t = x_{t-1} + \mu + \varepsilon_t, \qquad \varepsilon_t \sim \mathcal{N}(0, \sigma^2)$$

Recover the per-step drift $\mu$ and step noise $\sigma$.

Warning: everything you did in cases 1–2 assumed independent errors around a deterministic curve. This series is **not** that — each value carries the accumulated sum of all past noise. Fit a line to it directly and your standard errors will be fiction. There is a one-line transformation that turns this case back into case 1.`,
  recover: 'Recover drift μ and step std σ (per step).',
  starterCode: String.raw`# GROUND TRUTH · CASE 05 — The Drunkard's Drift
x = data.y.to_numpy()
t = data.t.to_numpy()

# the naive move (try it, watch the residuals lie to you):
b, a = np.polyfit(t, x, 1)
overlay_fn(lambda t: a + b*t, label="naive OLS line", style="dot")
residuals(a + b*t)

# the residuals look small. are they telling the truth?
`,
  hints: {
    nudge: String.raw`The *increments* $x_t - x_{t-1}$ are iid $\mathcal{N}(\mu, \sigma^2)$. Difference first; then you are back on case 1.`,
    method: String.raw`\`d = np.diff(x)\`; then $\hat\mu = $ \`d.mean()\` and $\hat\sigma = $ \`d.std(ddof=1)\`. The residuals of the naive line fit look small because a random walk *resembles* a trend on any single realization — the ACF of those residuals betrays the memory.`,
    derivation: String.raw`Differencing gives $d_t = x_t - x_{t-1} = \mu + \varepsilon_t$, iid — so

$$\hat\mu = \bar{d}, \quad \mathrm{SE}(\hat\mu) = \frac{\sigma}{\sqrt{N}} \approx 0.032; \qquad \hat\sigma^2 = \mathrm{var}(d), \quad \mathrm{SE}(\hat\sigma) \approx \frac{\sigma}{\sqrt{2N}} \approx 0.022$$

Note the asymmetry the brief promised: with $\sigma = 1$ and $\mu = 0.15$, the drift's own SE is a fifth of its value — barely a 6-sigma measurement after 1000 steps — while $\sigma$ is pinned to 2%. **Volatility is easy; drift is hard.** (Case 7 sharpens this into a theorem.)

Also note: $\hat\mu = \bar d = (x_N - x_0)/N$ — for a pure random walk the endpoints *are* the sufficient statistic for drift. The middle of the path adds nothing. Compare case 2, where endpoints-only was the worst thing you could do: whether the noise accumulates changes everything.`,
  },
  xLabel: 't (steps)',
  yLabel: 'x',
};

export default meta;

File: datagen/generators/level05.py
"""Level 5 — The Drunkard's Drift. Random walk with drift.

Estimator: difference first. Increments are iid N(mu, sigma^2), so
mu_hat = mean(diff), sigma_hat = std(diff). sigma is easy (SE ~ sigma/sqrt(2N));
mu is hard (SE = sigma/sqrt(N) regardless of how long you watch per-step).
"""

import numpy as np

from datagen.common import field, make_pack, rng, truth_block

SEED = 1505
N = 1000
MU = 0.15
SIGMA = 1.0
X0 = 20.0


def build() -> dict:
    g = rng(SEED)
    steps = MU + g.normal(0.0, SIGMA, N)
    x = X0 + np.concatenate([[0.0], np.cumsum(steps)])
    t = np.arange(N + 1, dtype=float)

    se_mu = SIGMA / np.sqrt(N)
    se_sigma = SIGMA / np.sqrt(2 * N)
    return make_pack(
        id=5,
        seed=SEED,
        kind="series",
        columns={"t": t, "y": x},
        truth=truth_block(
            {"mu": MU, "sigma": SIGMA},
            r"x_t = x_{t-1} + \mu + \varepsilon_t,\quad \mu = 0.15,\ \varepsilon_t \sim \mathcal{N}(0,\,1)",
            curve_t=t,
            curve_y=X0 + MU * t,
        ),
        grading={
            "mode": "parameter",
            "fields": [
                field("mu", "drift μ (per step)", MU, 4 * se_mu),
                field("sigma", "step std σ", SIGMA, 4 * se_sigma),
            ],
        },
    )
